crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
+/// Representation of an exclusive clone of an upstream repo
+///
+/// [UpstreamRepo] and [UpstreamCatalog] ensure safe access to individual catalog branches.
+/// Every [UpstreamRepo] instance represents an exclusive clone
+/// and can only ever create a single [UpstreamCatalog] instance at a time.
+struct UpstreamRepo(Git);
+
+impl UpstreamRepo {
+    /// Clone the upstream repo
+    async fn clone_repo(
+        publish_ref: &PublishRef,
+        temp_dir: impl AsRef<Path>,
+    ) -> Result<Self, PublishError> {
+        let url = publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(&url, &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
+    async fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError> {
+        if self.0.list_branches().await? // todo: catch error
+            .into_iter().any(|info| info.name == Self::catalog_branch_name(system))
+        {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), false)
+                .await?; // todo: catch error
+        } else {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), true)
+                .await?;
+            self.0
+                .set_origin(&Self::catalog_branch_name(system), "origin")
+                .await?;
+        }
+        Ok(UpstreamCatalog(&self.0))
+    }
+}
+
+/// Representation of a specific catalog branch in an exclusive clone of an upstream repo.
+///
+/// [UpstreamCatalog] guaranteesd that during its lifetime all operations on the underlying git repo
+/// are performed on a single branch,
+/// and that the branch is pushed to upstream before a new branch can be checked out.
+struct UpstreamCatalog<'a>(&'a Git);
+
+impl UpstreamCatalog<'_> {
+    /// Mostly naïve approxiaton of a snapshot path
+    /// TODO: fix before releasing this PR!
+    fn get_snapshot_path(&self, snapshot: &Value) -> PathBuf {
+        let mut path = self
+            .0
+            .workdir()
+            .unwrap()
+            .join("packages")
+            .join(snapshot["eval"]["meta"]["pname"].to_string());
+        snapshot["eval"]["meta"]["version"].to_string();
+
+        path.set_file_name(format!("{}.json", snapshot["eval"]["meta"]["version"]));
+        path
+    }
+
+    /// Try retrieving a snapshot from the catalog
+    /// TODO: better addressing (attrpath + version + drv hash?)
+    fn get_snapshot(&self, snapshot: &Value) -> Result<Option<Value>, PublishError> {
```

Comment:
I'm guessing you're hoping to replace `Value` with `CatalogEntry` throughout?

Reply:
The result, yes
the address (`snapshot: &Value`) I'd like to more lightweight

Reply:
But you'd expect to replace `snapshot["eval"]["meta"]["pname"].to_string()` above at some point?

Reply:
Yes the current `get_snapshot_path` is very preliminary and subject to change based on what we decide for the catalog to be like.